EIPS/eip-225.md

```diff
@@ -0,0 +1,474 @@
+
+**The protocol of maintaining the list of authorized signers must be fully contained in the block headers.**
+
+The next obvious idea would be to change the structure of the block headers so it drops the notions of PoW, and introduces new fields to cater for voting mechanisms. This is also the wrong answer: changing such a core data structure in multiple implementations would be a nightmare development, maintenance and security wise.
+
+**The protocol of maintaining the list of authorized signers must fit fully into the current data models.**
+
+So, according to the above, we can't use the EVM for voting, rather have to resort to headers. And we can't change header fields, rather have to resort to the currently available ones. Not much wiggle room.
+
+### Repurposing header fields for signing and voting
+
+The most obvious field that currently is used solely as *fun metadata* is the 32 byte **extra-data** section in block headers. Miners usually place their client and version in there, but some fill it with alternative "messages". The protocol would extend this field ~~to~~ with 65 bytes with the purpose of a secp256k1 miner signature. This would allow anyone obtaining a block to verify it against a list of authorized signers. It also makes the **miner** section in block headers obsolete (since the address can be derived from the signature).
+
+*Note, changing the length of a header field is a non invasive operation as all code (such as RLP encoding, hashing) is agnostic to that, so clients wouldn't need custom logic.*
+
+The above is enough to validate a chain, but how can we update a dynamic list of signers. The answer is that we can repurpose the newly obsoleted **miner** field and the PoA obsoleted **nonce** field to create a voting protocol:
+
+ * During regular blocks, both of these fields would be set to zero.
+ * If a signer wishes to enact a change to the list of authorized signers, it will:
+   * Set the **miner** to the signer it wishes to vote about
+   * Set the **nonce** to `0` or `0xff...f` to vote in favor of adding or kicking out
+
+Any clients syncing the chain can "tally" up the votes during block processing, and maintain a dynamically changing list of authorized signers by popular vote.
+
+To avoid having an infinite window to tally up votes in, and also to allow periodically flushing stale proposals, we can reuse the concept of an epoch from ethash, where every epoch transition flushes all pending votes. Furthermore, these epoch transitions can also act as stateless checkpoints containing the list of current authorized signers within the header extra-data. This permits clients to sync up based only on a checkpoint hash without having to replay all the voting that was done on the chain up to that point. It also allows the genesis header to fully define the chain, containing the list of initial signers.
```

Comment:
Is a proposal essentially a potential signer that may be added to list of authorized signers? It may be nice to have this defined too since it is a frequently term which may be confused with other concepts.

Reply:
It's detailed in the spec:

---

For all non-epoch transition blocks:

 * Signers may cast one vote per own block to propose a change to the authorization list.
 * Only the latest proposal per target beneficiary is kept from a single signer.
 * Votes are tallied live as the chain progresses (concurrent proposals allowed).
 * Proposals reaching majority consensus **`SIGNER_LIMIT`** come into effect immediately.
 * Invalid proposals are **not** to be penalized for client implementation simplicity.